### PR TITLE
Bugfix/custom localdir with user: Fixes #462

### LINKED ISSUE #462: Create jenkins user before installing the package with manage_user
The usermod fails when using a custom localestatedir value due to jenkins already running after installation (debian-ish systems are 'cool' like that)

```
==> jenkins: Notice: /Stage[main]/Jenkins::Package/Package[jenkins]/ensure: ensure changed 'purged' to 'present'
==> jenkins: Error: Could not set home on user[jenkins]: Execution of '/usr/sbin/usermod -d /data/jenkins jenkins' returned 8: usermod: user jenkins is currently used by process 17664
==> jenkins: Error: /Stage[main]/Jenkins::Config/User[jenkins]/home: change from /var/lib/jenkins to /data/jenkins failed: Could not set home on user[jenkins]: Execution of '/usr/sbin/usermod -d /data/jenkins jenkins' returned 8: usermod: user jenkins is currently used by process 17664
```

### PR DESCRIPTION
**Pull request by vStone**
By splitting out the user creation, we can have a custom
localdir before the package (maybe) tries to create it or adjust it.

### CODE REVIEW

**Comment by vStone** [comment]
Looks like the tests fail due to the repo class having support for generic linux but the sysconfig define does not.

**Comment by vStone** [comment]
@jhoblitt / @rtyler: Do we really want to assume that a unknown osfamily is a redhat variant? If we do, we should probably assume this for all `case $::osfamily` in the code...

**Comment by jhoblitt** [comment]
Probably not.  The intent of #413 is to audit and cleanup platform handling.

**Comment by jhoblitt** [comment]
I'm not enthusiastic about the name `pre_config`, as it is fairly ambiguous.  How about something like 'user_setup` or `role_account`?

File: manifests/config.pp [hunk 1]
```diff
@@ -7,37 +7,10 @@
...
-  # old conditional behavior of jenkins::plugin
-  if $::jenkins::manage_user {
-    ensure_resource('user', $::jenkins::user, {
-      ensure     => present,
-      gid        => $::jenkins::group,
-      home       => $::jenkins::localstatedir,
-      managehome => false,
-      system     => true,
-    })
+  if ! has_key($::jenkins::config_hash, 'JENKINS_HOME') {
```
Comment (jhoblitt) [inline manifests/config.pp]: If a `Jenkins::Sysconfig['JENKINS_HOME']` resource is already defined, this will fail.  I don't think it is safe to assume that it won't happen as `jenkins::sysconfig` is a public interface (I'd rather we didn't have both `config_hash` and the defined type but we are stuck for the moment). 

I think using `defined` to check for the existence of the resource might be marginally safer?  That doesn't prevent the end user from causing a duplicate resource declaration by having a `JENKINS_HOME` key in `config_hash` and a `jenkins::sysconfig` resource but I feel it should be more obvious to the end user that they shouldn't define the same value twice.

**Comment by vStone** [comment]
I've renamed the class and in stead of only checking for either the resource in the hash or defined as a recourse, I do both.

**Comment by vStone** [comment]
Rebased and attempt to fix some facter osfamily vs os['distro']['family']-ish issues. Needs a better solution in another pull request.

**Comment by jhoblitt** [comment]
I'm afraid this isn't safe to merge because of the puppet 4.x breakage.

**Comment by vStone** [comment]
I've put it in the same state as any other build. The beaker/docker builds fail, but everything else is green.

**Comment by vStone** [comment]
@jhoblitt any remarks?

**Comment by jhoblitt** [comment]
I've kicked travis.  I'm a bit nervous that all 3 acceptance tests are failing.

**Comment by vStone** [comment]
```The command "bundle exec rake parallel_spec SPEC_OPTS='--format documentation --order random'" exited with 0.

Done. Your build exited with 1.
```
o.O

**Comment by vStone** [comment]
Rebased on master.

**Comment by vStone** [comment]
@jhoblitt Care to give this another set of eyes?

I rebased on master and all travis builds, except the one that should not be building (puppet 3.8.0 on ruby 2.2) are currently in the green.

File: manifests/config.pp [hunk 2]
```diff
@@ -14,36 +14,13 @@
...
   create_resources('jenkins::sysconfig', $config_hash)
 
-  $dir_params = {
-    ensure => directory,
-    owner  => $::jenkins::user,
-    group  => $::jenkins::group,
-    mode   => '0755',
+  if (
+    ! defined(Jenkins::Sysconfig['JENKINS_HOME'])  and
+    ! has_key($::jenkins::config_hash, 'JENKINS_HOME')
```
Comment (jhoblitt) [inline manifests/config.pp]: Is this necessary?  This is asking for parse order problems.

File: spec/classes/jenkins_user_setup_spec.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  let (:facts) { {
+    :osfamily => 'RedHat',
+    :operatingsystem => 'RedHat',
+    :operatingsystemrelease => '7.4',
+    :operatingsystemmajrelease => '7'
+  } }
+
+  context 'user_setup' do
+    context 'default' do
+      it { should contain_class('jenkins::config') }
```
Comment (jhoblitt) [inline spec/classes/jenkins_user_setup_spec.rb]: This seems unrelated to what is being tested?  It is a private class too.

File: spec/classes/jenkins_user_setup_spec.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      end
+    end
+    context 'unmanaged' do
+      let (:params) {{
+        :manage_user     => false,
+        :manage_group    => false,
+        :manage_datadirs => false,
+      }}
+      it { should_not contain_user('jenkins') }
+      it { should_not contain_user('jenkins') }
```
Comment (jhoblitt) [inline spec/classes/jenkins_user_setup_spec.rb]: Duplicate statement.

File: spec/classes/jenkins_user_setup_spec.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      it { should_not contain_group('jenkins') }
+      it { should_not contain_file('/var/lib/jenkins') }
```
Comment (jhoblitt) [inline spec/classes/jenkins_user_setup_spec.rb]: What about `.../{jobs,plugins}`?

File: manifests/config.pp (hunk 2 above)
Comment (vStone) [inline manifests/config.pp]: Not checking will make sure manifests fail when they have a JENKINS_HOME in the config hash or manually defined. I should probably document the use case for having a different localstatedir from the jenkins's user home dir that will require manage_user => false then.

**Comment by vStone** [comment]
@jhoblitt: all remarks addressed.

**Comment by vStone** [comment]
@jhoblitt It took a while, but: rebased.

**Comment by vStone** [comment]
@jhoblitt request for re-review. Can we get this merged?

**Comment by vStone** [comment]
@jhoblitt please either merge this or decline the pull request (with a better solution) ;)

**Comment by vStone** [comment]
Now that we (voxpupuli) have ownership, can we re-discuss this @jhoblitt ? overriding the default location with a custom user is still a problem it seems. I'm open to alternative implementations to get this fixed. 

Note: a solution might be to disallow overriding the default /var/lib/jenkins location.

**Comment by jhoblitt** [comment]
@vStone Could you rebase on current master?  This looks mergable as long as travis passes.

File: manifests/user_setup.pp [hunk 6]
```diff
@@ -0,0 +1,53 @@
+# This class hould be considered private
+#
```
Comment (jhoblitt) [inline manifests/user_setup.pp]: It would be useful to have a blurb discussing why `user/group` management needs to be handled before the package is installed.

Comment (vStone) [reply manifests/user_setup.pp]: We had issues with using a custom home directory

Comment (jhoblitt) [reply manifests/user_setup.pp]: I meant a blurb as a comment on the class.

**Comment by bastelfreak** [comment]
I restarted the failed acceptance test

**Comment by vStone** [comment]
They seem to not want to work on travis. I did get a complete run on my machine  k(͡๏̯͡⊙)ゞ
